modules/routes/Forums/forum_route.py

```diff
@@ -0,0 +1,147 @@
+    session = fetch_session()
+    db_cursor = db.cursor(dictionary=True)
+    if session["account_type"] == AccountType.Student.name:
+        return jsonify({"message": "Students are not allowed to create forums"}), 403
+
+    # Check if the user is a lecturer who teaches this course
+    if session["account_type"] == AccountType.Lecturer.name:
+        # Check if the lecturer teaches the specified course
+        lecture_details = db_cursor.execute(
+            "SELECT * FROM Course WHERE lecturer_id = %s", (session["sub"],)
+        ).fetchone()
+        if course_code != lecture_details["course_code"]:
+            return jsonify({"message": "You can only view your own courses!"}), 403
+
+    # Proceed with retrieving discussion threads for the forum
+    db_cursor.execute("SELECT * FROM DiscussionThread WHERE forum_id = %s", (forum_id,))
+    threads = db_cursor.fetchall()
+    if not threads:
+        return jsonify({"message": "No discussion threads found for this forum"}), 404
+    # Convert the threads data into JSON format and return
+    return jsonify(threads), 200
+
+@app.route("/threads/forum/<int:forum_id>", methods=["POST"])
+@protected_route(roles=[AccountType.Admin, AccountType.Lecturer, AccountType.Student])
+def add_thread_to_forum(forum_id, course_code):
+    # Check if the user is a student
+    session = fetch_session()
+    db_cursor = db.cursor(dictionary=True)
+    if session["account_type"] == AccountType.Student.name:
+        return jsonify({"message": "Students are not allowed to create forums"}), 403
+
+    # Check if the user is a lecturer who teaches this course
+    if session["account_type"] == AccountType.Lecturer.name:
+        # Check if the lecturer teaches the specified course
+        lecture_details = db_cursor.execute(
+            "SELECT * FROM Course WHERE lecturer_id = %s", (session["sub"],)
+        ).fetchone()
+        if course_code != lecture_details["course_code"]:
+            return jsonify({"message": "You can only view your own courses!"}), 403
+            
+    # Proceed with adding the discussion thread to the forum
+    body = NewDiscussionThreadSchema().load(request.get_json(force=True))
+    user_id = session["user_id"]  # Assuming user_id is obtained from JWT token
+
+    try:
+        db_cursor.execute(
+            "INSERT INTO DiscussionThread (title, post, forum_id, user_id) VALUES (%s, %s, %s, %s)",
+            (body["title"], body["post"], forum_id, user_id),
+        )
+        db.commit()
+        return jsonify({"message": "Discussion thread added successfully"}), 201
+    except Exception as e:
+        db.rollback()
+        return jsonify({"message": f"Failed to add discussion thread: {str(e)}"}), 500
+
+
+@app.route("/thread/<int:thread_id>/reply", methods=["POST"])
+@protected_route()
+def add_reply_to_thread(thread_id):
+    body = NewDiscussionReplySchema().load(request.get_json(force=True))
+    user_id = fetch_session()["sub"]
+
```

Comment:
Where are the checks to see if a student is enrolled for the course or a lecturer teaches the course??